BaragonAgentService/src/main/java/com/hubspot/baragon/agent/managed/LifecycleHelper.java

```diff
@@ -205,4 +228,72 @@ public void shutdown() throws Exception {
       removeStateFile();
     }
   }
+
+  public void setBootstrapStateNodeVersion(Optional<Integer> version) {
+    if (version.isPresent()) {
+      bootstrapStateNodeVersion.set(version.get());
+    }
+  }
+
+  public int getBootstrapStateNodeVersion() {
+    return bootstrapStateNodeVersion.get();
+  }
+
+  public void checkStateNodeVersion() {
+    agentState.set(BaragonAgentState.BOOTSTRAPING);
+    try {
+      Optional<Integer> maybeStateVersion = stateDatastore.getStateVersion();
+      if (maybeStateVersion.isPresent()) {
+        if (!agentLock.tryLock(agentLockTimeoutMs, TimeUnit.MILLISECONDS)) {
+          throw new LockTimeoutException("Could not acquire lock to reapply configs", agentLock);
+        }
+        try {
+          if (getBootstrapStateNodeVersion() < maybeStateVersion.get()) {
+            applyCurrentConfigs();
+          }
+        } catch (Exception e) {
+          abort("Could not ensure configs are up to date, aborting", e);
+        } finally {
+          agentLock.unlock();
+        }
+      }
+    } catch (Exception e) {
+      abort("Interrupted while trying to reapply configs, shutting down", e);
+    }
+    agentState.set(BaragonAgentState.ACCEPTING);
+  }
+
+  @SuppressFBWarnings("DM_EXIT")
+  public void abort(String message, Exception exception) {
+    LOG.error(message, exception);
+    flushLogs();
+    Optional<Server> server = serverProvider.get();
+    if (server.isPresent()) {
+      try {
+        server.get().stop();
+        shutdown();
```

Comment:
shutdown() i know blocks, need to check on .stop() for the server. tbh I copied most of this from how singularity does its abort